be/src/exec/workgroup/scan_task_queue.h

```diff
@@ -61,65 +66,84 @@ class PriorityScanTaskQueue final : public ScanTaskQueue {
+    size_t size() const override { return _num_tasks.load(std::memory_order_acquire); }
 
-    void close() override;
+    void update_statistics(WorkGroup* wg, int64_t runtime_ns) override;
+    bool should_yield(const WorkGroup* wg, int64_t unaccounted_runtime_ns) const override;
 
-    StatusOr<ScanTask> take(int worker_id) override;
-    bool try_offer(ScanTask task) override;
+private:
+    /// These methods should be guarded by the outside _global_mutex.
+    workgroup::WorkGroupScanSchedEntity* _take_next_wg();
+    void _update_min_wg();
+    // Apply hard bandwidth control to non-realtime workgroups, when there are queries of the realtime workgroup.
+    bool _throttled(const workgroup::WorkGroupScanSchedEntity* wg_entity, int64_t unaccounted_runtime_ns = 0) const;
+    void _update_bandwidth_period();
 
-    size_t size() const override { return _total_task_num.load(std::memory_order_acquire); }
+    int64_t _bandwidth_quota_ns() const;
+    // The ideal runtime of a work group is the weighted average of the schedule period.
+    int64_t _ideal_runtime_ns(workgroup::WorkGroupScanSchedEntity* wg_entity) const;
 
 private:
-    // Calculate the actual cpu used by all wg
-    void _cal_wg_cpu_real_use_ratio();
+    static constexpr int64_t SCHEDULE_PERIOD_PER_WG_NS = 100'000'000;
+    static constexpr int64_t BANDWIDTH_PERIOD_NS = 100'000'000;
 
-    // _maybe_adjust_weight and _select_next_wg are guarded by the ourside _global_mutex.
-    void _maybe_adjust_weight();
-    WorkGroupPtr _select_next_wg(int worker_id);
+    struct WorkGroupScanSchedEntityComparator {
+        using WorkGroupScanSchedEntityPtr = workgroup::WorkGroupScanSchedEntity*;
+        bool operator()(const WorkGroupScanSchedEntityPtr& lhs, const WorkGroupScanSchedEntityPtr& rhs) const;
+    };
+    using WorkgroupSet = std::set<workgroup::WorkGroupScanSchedEntity*, WorkGroupScanSchedEntityComparator>;
 
-    static constexpr int MAX_SCHEDULE_NUM_PERIOD = 512;
-
-    std::mutex _global_mutex;
+    mutable std::mutex _global_mutex;
     std::condition_variable _cv;
-
-    const ScanExecutorType _type;
     bool _is_closed = false;
 
-    std::unordered_set<WorkGroupPtr> _ready_wgs;
-    std::atomic<size_t> _total_task_num = 0;
-
-    // Adjust select factor of each wg after every `min(MAX_SCHEDULE_NUM_PERIOD, num_tasks)` schedule times.
-    int _remaining_schedule_num_period = 0;
+    // Contains the workgroups which include the drivers ready to be run.
+    // The entity is sorted by vruntime in set.
+    // MUST guarantee the entity is not in set, when updating its vruntime.
+    WorkgroupSet _ready_wg_entities;
+
+    size_t _sum_cpu_limit = 0;
+
+    // Cache the minimum vruntime and entity, used to check should_yield() without lock.
+    std::atomic<int64_t> _min_vruntime_ns = std::numeric_limits<int64_t>::max();
+    std::atomic<workgroup::WorkGroupScanSchedEntity*> _min_wg_entity = nullptr;
+
+    // Hard bandwidth control to non-realtime workgroups.
```

Comment:
The word `bandwidth` is not so precise? Maybe `time-slice`, or some other words?

Reply:
bandwidth control is from linux kernel hard limit for cgroup.